src/shell_history.rs

```diff
@@ -49,80 +52,161 @@ pub fn history_file_path() -> PathBuf {
     })
 }
 
-pub fn full_history(path: &PathBuf) -> Vec<String> {
-    let history_contents = read_ignoring_utf_errors(&path);
+/// Represents each entry in a history file.
+#[derive(Debug)]
+pub struct HistoryCommand {
+    /// The user's command.
+    pub command: String,
+    /// When the command was run, in seconds since Unix epoch.
+    pub when: i64,
+    /// The format of the file, so we can write the record back out.
+    pub format: HistoryFormat,
+}
 
-    let zsh_timestamp_and_duration_regex = Regex::new(r"^: \d+:\d+;").unwrap();
+impl HistoryCommand {
+    pub fn new<S>(command: S, when: i64, format: HistoryFormat) -> Self
+    where
+        S: Into<String>,
+    {
+        Self {
+            command: command.into(),
+            when,
+            format,
+        }
+    }
+}
+
+impl fmt::Display for HistoryCommand {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.format {
+            HistoryFormat::Plain {
+                zsh_extended_history,
+            } => {
+                if zsh_extended_history {
+                    write!(f, ": {}:0;{}", self.when, self.command)
+                } else {
+                    write!(f, "{}", self.command)
+                }
+            }
+            HistoryFormat::Fish => writeln!(f, "- cmd: {}\n  when: {}", self.command, self.when),
+        }
+    }
+}
 
-    history_contents
-        .split('\n')
-        .filter(|line| !has_leading_timestamp(line) && !line.is_empty())
-        .map(|line| zsh_timestamp_and_duration_regex.replace(line, ""))
-        .map(String::from)
-        .collect::<Vec<String>>()
+pub fn full_history(path: &PathBuf, history_format: HistoryFormat) -> Vec<HistoryCommand> {
+    match history_format {
+        HistoryFormat::Plain { .. } => {
+            let history_contents = read_ignoring_utf_errors(&path);
+            let zsh_timestamp_and_duration_regex = Regex::new(r"^: \d+:\d+;").unwrap();
+            let when = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or_else(|err| panic!("McFly error: Time went backwards ({})", err))
+                .as_secs() as i64;
+            history_contents
+                .split('\n')
+                .filter(|line| !has_leading_timestamp(line) && !line.is_empty())
+                .map(|line| zsh_timestamp_and_duration_regex.replace(line, ""))
```

Comment:
You've set this up nicely to actually use the ZSH timestamp if it exists, which we're currently throwing away. :+1: